### ISSUE TITLE
Missing events: Not getting any event for the recreated directory

### ISSUE THREAD

**Issue by vivek-bansal-VB** [rjeczalik/notify#190]
I have set a watch on directory /home/mydir and created a subdirectory (tmp) under it. I dropped some files in directory /home/mydir/tmp and I got the file create events. Then I deleted the directory (/home/mydir/tmp) and recreated it. Again I dropped some files in the directory (/home/mydir/tmp), but this time I am not getting any events in this directory. Seems a probable bug. Any help is much appreciated.

**Comment by rjeczalik** [comment]
Could you please share some reproducer? Go snippet to show how you set watches and some example bash commands how you create temp dir and do all these operations. It would be greatly helpful.

**Comment by vivek-bansal-VB** [comment]
`func main() {
	WatchFS("/home/mydir")
}

func WatchFS(rootDir string64) {
	events := make(chan notify.EventInfo, 100)
	setupWatch(rootDir, events)
	handleFSEventsInfinite(events, rootDir)
	notify.Stop(fsEvents)
}

func setupWatch(path string, events chan notify.EventInfo) error {
	if err := notify.Watch(path, events, notify.All, notify.Create, notify.InCreate); err != nil {
		fmt.Errorf("Failed to setup watch on %q : %v.", path, err)
		return err
	}
	fmt.Infof("Setup watch on %q", path)
	return nil
}

func handleFSEventsInfinite(events chan notify.EventInfo) {
	for {
		ei := <-events
		switch event := ei.Event(); event {
		case notify.Create, notify.InCreate:
			if fi.IsDir() {
				fmt.Infof("Creation event is a directory %q", path)
				setupWatch(path, events)
			} else {
				fmt.Infof("event is a file create event on file: %q", path)
			}
		default:
			fmt.Infof("Got non-create event %q", event)
		}
	}
}`

**Comment by vivek-bansal-VB** [comment]
Below is the shell script which I used to run the commands:
`sudo mkdir /home/mydir/tmp
cd /home/mydir/tmp
touch a.txt
cd ..
sudo rm -rf tmp
sudo mkdir /home/mydir/tmp
touch b.txt`

**Comment by vivek-bansal-VB** [comment]
Output of the above script:
event is a directory create event: "/home/mydir/tmp"
event is a directory create event: "/home/mydir/tmp"
event is a directory create event: "/home/mydir/tmp"
event is a file create event on file: "/home/mydir/tmp/a.txt"
event is a file create event on file: "/home/mydir/tmp/a.txt"
Got non-create event "notify.Remove"
Got non-create event "notify.Remove"
Got non-create event "notify.Remove"
Got non-create event "notify.Remove"
Got non-create event "notify.Remove"
event is a directory create event: "/home/mydir/tmp"
event is a directory create event: "/home/mydir/tmp"
event is a directory create event: "/home/mydir/tmp"
event is a directory create event: "/home/mydir/tmp"

**Comment by vivek-bansal-VB** [comment]
As you can see, there is no event generated for file /home/mydir/tmp/b.txt

**Comment by vivek-bansal-VB** [comment]
I checked this package: https://github.com/fsnotify/fsnotify and it seems to be working fine even when the directory with the same name is recreated.

**Comment by rjeczalik** [comment]
@vivek-bansal-VB Thanks for the repro! Since I don't have access to macOS nowadays, someone else would need to troubleshoot it.

Out of curiosity, does it reproduce when you build with kqueue instead of FSEvents? `go build -tags kqueue ...`

**Comment by vivek-bansal-VB** [comment]
@rjeczalik Yes I tried with -tags kqueue and the same issue is there as well.

**Comment by vivek-bansal-VB** [comment]
Other solution for this problem can be: If there would be a function: RemoveWatch(path) exposed by the library. So before deleting the directory I will remove the watch from that directory, and later on when the same directory will be created again, AddWatch(path) method will work on that correctly. Currently AddWatch(path) method is not working correctly since I didn't remove the watch from the directory as there is no method exposed to remove the watch.

**Comment by stanvarlamov** [comment]
Same exact issue on Linux/Ubuntu 18.04.5 LTS. A go routine runs cycles that create two subfolders in the watched dir via `MkdirAll` and write a file into the lower one: `/watched-dir/123/456/test.abc` then `rm -r /watched-dir/123`; pauses for 3 sec after writes and deletes. The first and second cycles show all objects' events displayed in the watcher properly. From the 3rd cycle on, only events for `/watched-dir/123` are showing in the watcher. So it appears that the watcher looses its recursivity. I haven't tried resetting the watcher as per the last comment as that kind of defeats the purpose of using a recursive watcher: if I have to trace dir events in the path I may as well just switch to a set of non-recursive watchers.   @rjeczalik - when do you think you can look at this? Pls let me know if you'd like the tester snipped to be posted

**Comment by imsodin** [comment]
Might also be fixed by #201 - could you test if it actually is please?  @vivek-bansal-VB
